python/pyspark/logger/logger.py

```diff
@@ -0,0 +1,150 @@
+    - timestamp: The time the log record was created.
+    - level: The log level of the record.
+    - name: The name of the logger.
+    - message: The log message.
+    - kwargs: Any additional keyword arguments passed to the logger.
+    """
+
+    def format(self, record: logging.LogRecord) -> str:
+        """
+        Format the specified record as a JSON string.
+
+        Parameters
+        ----------
+        record : logging.LogRecord
+            The log record to be formatted.
+
+        Returns
+        -------
+        str
+            The formatted log record as a JSON string.
+        """
+        log_entry = {
+            "timestamp": self.formatTime(record, self.datefmt),
+            "level": record.levelname,
+            "name": record.name,
+            "message": record.getMessage(),
+        }
+        log_entry.update(record.__dict__.get("kwargs", {}))
+        return json.dumps(log_entry, ensure_ascii=False)
+
+
+class PySparkLogger(logging.Logger):
+    """
+    Custom logger for PySpark that logs messages in a structured JSON format.
+
+    This logger provides methods for logging messages at different levels (info, warning, error)
+    with additional keyword arguments that are included in the JSON log entry.
+    """
+
+    def __init__(
+        self,
+        name: str = "PySparkLogger",
+        level: int = logging.INFO,
+        stream: Optional[Union[None, io.TextIOWrapper]] = None,
+        filename: Optional[str] = None,
+    ):
+        super().__init__(name, level)
+        if not self.hasHandlers():
+            self.handler = (
+                logging.FileHandler(filename) if filename else logging.StreamHandler(stream)
+            )
+            self.handler.setFormatter(JSONFormatter())
+            self.addHandler(self.handler)
+            self.setLevel(level)
+
+    @staticmethod
+    def get_logger(
+        name: str = "PySparkLogger",
+        level: int = logging.INFO,
+        stream: Optional[Union[None, io.TextIOWrapper]] = None,
+        filename: Optional[str] = None,
+    ) -> "PySparkLogger":
+        """
+        Get a configured instance of PySparkLogger.
+
+        Parameters
+        ----------
+        name : str, optional
+            The name of the logger (default is "PySparkLogger").
+        level : int, optional
+            The log level of the logger (default is logging.INFO).
+        stream : Optional[Union[None, io.TextIOWrapper]], optional
+            The stream to which log messages are written (default is None, which uses sys.stderr).
+        filename : Optional[str], optional
+            The name of the file to which log messages are written (default is None).
+
+        Returns
+        -------
+        PySparkLogger
+            A configured instance of PySparkLogger.
+        """
+        return PySparkLogger(name, level, stream, filename)
+
+    def log_info(self, message: str, **kwargs: Any) -> None:
```

Comment:
Is there a reason to add non idiomatic methods here? Python users will just know the methods to use that come from the logger?

Reply:
The reason for introducing `log_info`, `log_warn`, and `log_error`, instead of using the conventional `info`, `warn`, and `error`, is for consistency with Scala's `logInfo`, `logWarn`, and `logError` methods.

I believe this helps developers who work with both languages to have a consistent experience - I believe this also helps differentiate between Python's built-in `logging` and `PySparkLogger`, preventing conflicts -.

Please let me know if this doesn't fully address your question. Thanks! :-)

Reply:
So the tricky part here is the following:

* We have a Scala precedence that users might know,
* We add similarly named methods to the logger object,
* But instead of naming them exactly like the Scala ones, we opt to change them to a pythonic Snake case which renders the first point mute.

For the second part, I think the PySpark logger should absolutely follow the interface of the standard built in logger so that the Python developers that are familiar with it can use it well.

Reply:
Makes sense to me. Let me update according to your suggestion. Thanks!!